providers/yaml/src/main/java/org/jboss/resteasy/plugins/providers/YamlProvider.java

```diff
@@ -110,4 +143,131 @@ public void writeTo(Object t, Class<?> type, Type genericType, Annotation[] anno
+         if (configuration != null) {
+            value = configuration.getParameter(key);
+            if (value == null) {
+               value = configuration.getInitParameter(key);
+            }
+         }
+         if (value == null) {
+            System.getProperty(key);
+         }
+         return value;
+      } else {
+         return AccessController.doPrivileged((PrivilegedAction<String>) () -> {
+            String value = null;
+            if (configuration != null) {
+               value = configuration.getParameter(key);
+               if (value == null) {
+                  value = configuration.getInitParameter(key);
+               }
+            }
+            if (value == null) {
+               System.getProperty(key);
+            }
+            return value;
+         });
+      }
+   }
+
+   private static class TypeSafeConstructor extends CustomClassLoaderConstructor {
+      private final Set<Class<?>> types;
+      private final Collection<String> allowedList;
+
+      private TypeSafeConstructor(final ParameterizedType parameterizedType, final ClassLoader classLoader,
+                                  final Collection<String> allowedList) {
+         super(classLoader);
+         this.allowedList = allowedList;
+         final Set<Class<?>> genericTypes = new HashSet<>();
+         for (Type typeArg : parameterizedType.getActualTypeArguments()) {
+            try {
+               genericTypes.add(classLoader.loadClass(typeArg.getTypeName()));
+            } catch (ClassNotFoundException e) {
+               LogMessages.LOGGER.failedToLoadType(typeArg.getTypeName());
+            }
+         }
+         this.types = genericTypes;
+      }
+
+      private TypeSafeConstructor(final Class<?> type, final ClassLoader classLoader,
+                                  final Collection<String> allowedList) {
+         super(classLoader);
+         this.types = Collections.singleton(type);
+         this.allowedList = allowedList;
+      }
+
+      @Override
+      protected Object newInstance(final Class<?> ancestor, final Node node, final boolean tryDefault) {
+         if (denied(node.getType())) {
+            throw Messages.MESSAGES.typeNotAllowed(node.getType());
+         }
+         return super.newInstance(ancestor, node, tryDefault);
+      }
+
+      private boolean denied(final Class<?> type) {
+         if (type == null) {
+            return false;
+         }
+         final String name = type.getName();
+         // Check the allowed types first
+         if (allowedList.contains(name)) {
+            return false;
+         }
+         // Allow all primitives
+         if (type.isPrimitive()) {
+            return false;
+         }
+         // Denied types are not safe
+         if (DENY_LIST.contains(name)) {
+            return true;
+         }
+         // We should trust known types
+         if (name.startsWith("java") || name.startsWith("javax")) {
+            return false;
+         }

```

Comment:
Here's a payload to bypass this check:

```yaml
!!java.rmi.server.UnicastRemoteObject [6666]
```

Reply:
This is why deny lists are not great :) However, creating an allow list with every Java type is also not ideal. TBH we should probably deny any `java.rmi` type. I see no reason why that would be a generic type.